os/drivers/audio/syu645b.c

```diff
@@ -0,0 +1,1055 @@
+
+		priv->paused = true;
+		priv->mute = true;
+#ifndef CONFIG_AUDIO_EXCLUDE_VOLUME
+		syu645b_setvolume(priv);
+#endif
+		I2S_PAUSE(priv->i2s, I2S_TX);
+	}
+
+	syu645b_givesem(&priv->devsem);
+
+	return OK;
+}
+#endif							/* CONFIG_AUDIO_EXCLUDE_PAUSE_RESUME */
+
+/****************************************************************************
+ * Name: syu645b_resume
+ *
+ * Description: Resumes the playback.
+ *
+ ****************************************************************************/
+#ifndef CONFIG_AUDIO_EXCLUDE_PAUSE_RESUME
+#ifdef CONFIG_AUDIO_MULTI_SESSION
+static int syu645b_resume(FAR struct audio_lowerhalf_s *dev, FAR void *session)
+#else
+static int syu645b_resume(FAR struct audio_lowerhalf_s *dev)
+#endif
+{
+	FAR struct syu645b_dev_s *priv = (FAR struct syu645b_dev_s *)dev;
+
+	if (!priv) {
+		return -EINVAL;
+	}
+
+	audvdbg(" syu645b_resume Entry\n");
+	syu645b_takesem(&priv->devsem);
+
+	if (priv->running && priv->paused) {
+		priv->paused = false;
+		priv->mute = false;
+#ifndef CONFIG_AUDIO_EXCLUDE_VOLUME
+		syu645b_setvolume(priv);
+#endif
+
+		I2S_RESUME(priv->i2s, I2S_TX);
+
+	}
+
+	syu645b_givesem(&priv->devsem);
+	return OK;
+}
+#endif							/* CONFIG_AUDIO_EXCLUDE_PAUSE_RESUME */
+
+/****************************************************************************
+ * Name: syu645b_rxcallback
+ *
+ * Description: Called when I2S filled a buffer. No recycling mechanism now.
+ *
+ ****************************************************************************/
+
+static void syu645b_txcallback(FAR struct i2s_dev_s *dev, FAR struct ap_buffer_s *apb, FAR void *arg, int result)
+{
+	FAR struct syu645b_dev_s *priv = (FAR struct syu645b_dev_s *)arg;
+
+	DEBUGASSERT(priv && apb);
+
+	syu645b_takesem(&priv->devsem);
+	sq_entry_t *tmp;
+	for (tmp = (sq_entry_t *)sq_peek(&priv->pendq); tmp; tmp = sq_next(tmp)) {
+		if (tmp == (sq_entry_t *)apb) {
+			sq_rem(tmp, &priv->pendq);
+			audvdbg("found the apb to remove 0x%x\n", tmp);
+			break;
+		}
+	}
+
+	/* Call upper callback, let it post msg to user q */
+	priv->dev.upper(priv->dev.priv, AUDIO_CALLBACK_DEQUEUE, apb, OK);
+
+	syu645b_givesem(&priv->devsem);
+}
+
+/****************************************************************************
+ * Name: syu645b_enqueuebuffer
+ *
+ * Description: Enqueue an Audio Pipeline Buffer for playback/ processing.
+ *
+ ****************************************************************************/
+static int syu645b_enqueuebuffer(FAR struct audio_lowerhalf_s *dev, FAR struct ap_buffer_s *apb)
+{
+	FAR struct syu645b_dev_s *priv = (FAR struct syu645b_dev_s *)dev;
+	int ret;
+
+	if (!priv || !apb) {
+		return -EINVAL;
+	}
+
+	if (!priv->running) {
+		/* Add the new buffer to the tail of pending audio buffers */
+		syu645b_takesem(&priv->devsem);
+		sq_addlast((sq_entry_t *)&apb->dq_entry, &priv->pendq);
+		syu645b_givesem(&priv->devsem);
+		return OK;
+	}
+	
+	ret = I2S_SEND(priv->i2s, apb, syu645b_txcallback, priv, SYU645B_I2S_TIMEOUT_MS);
```

Comment:
Last parameter, actually it should be calculated based on samplerate & bitrate & buffer size.
For example, if sample rate is 48k and bitreate is 16bit, and buffer size is 8192, then 12 times of i2s should be executed within 1s.
In this case timeout value should be set as 80~90 ms.

Reply:
@ritesh55555 did you check it?

Reply:
updated

Reply:
I mean it should be caluclated dynamically, not fixed value.